sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/cardinality/ApproximateDistinct.java

```diff
@@ -0,0 +1,352 @@
+   * thus the number of different "buckets" in the HyperLogLog+ sketch.
+   * <br>In general, you can expect a relative error of about :
+   * <pre>{@code 1.1 / sqrt(2^p)}</pre>
+   * For instance, the estimation {@code ApproximateDistinct.globally(12)}
+   * will have a relative error of about 2%.
+   * <br> Also keep in mind that {@code p} cannot be lower than 4,
+   * because the estimation would be too inaccurate.
+   * <br>See {@link ApproximateDistinctFn} for more details about the algorithm's principle.
+   *
+   * <p>HyperLogLog+ version of Google uses a sparse representation in order to
+   * optimize memory and improve accuracy for small cardinalities.
+   * By calling this builder, you will not use the sparse representation.
+   * If you want to, see {@link ApproximateDistinctFn#withSparseRepresentation(int)}
+   *
+   * <p>Example of use
+   * <pre> {@code PCollection<KV<Integer, String>> input = ...;
+   * PCollection<KV<Integer, HyperLogLogPlus>> hllSketch = input
+   *        .apply(ApproximateDistinct.<Integer, String>perKey(15));
+   * } </pre>
+   *
+   * @param p       number of bits for indexes in the HyperLogLogPlus.
+   *
+   */
+  public static <K, InputT> Combine.PerKey<K, InputT, HyperLogLogPlus> perKey(
+      int p) {
+    return Combine.<K, InputT, HyperLogLogPlus>perKey(ApproximateDistinctFn.<InputT>create(p));
+  }
+
+    /**
+     * Do the same as {@link ApproximateDistinct#globally(int)}, but with a default value for p.
+     *
+     * @param <K>       the type of the keys in the input and output {@code PCollection}s
+     * @param <InputT>  the type of values in the input {@code PCollection}
+     */
+  public static <K, InputT> Combine.PerKey<K, InputT, HyperLogLogPlus> perKey() {
+    return perKey(18);
+  }
+
+  /**
+   * A {@code Combine.CombineFn} that computes the stream into a {@link HyperLogLogPlus}
+   * sketch, useful as an argument to {@link Combine#globally} or {@link Combine#perKey}.
+   *
+   * <p>The HyperLogLog algorithm relies on the principle that the overall cardinality
+   * can be estimated thanks to the longest run of starting 0s of the hashed elements.
+   * The longer the run is, the more unlikely it was to happen. Thus, the greater the number
+   * of elements that have been hashed before getting such a run.
+   * <br>Because this algorithm relies mainly on randomness, the stream is divided into buckets
+   * in order to reduce the variance of the estimation.
+   * Therefore, an estimation is applied on several samples and the overall estimation
+   * is then computed using an average (Harmonic mean).
+   *
+   * @param <InputT>    the type of the elements being combined
+   */
+  public static class ApproximateDistinctFn<InputT>
+      extends Combine.CombineFn<InputT, HyperLogLogPlus, HyperLogLogPlus> {
+
+    private final int p;
+
+    private final int sp;
+
+    private ApproximateDistinctFn(int p, int sp) {
+      this.p = p;
+      this.sp = sp;
+    }
+
+      /**
```

Comment:
<!--new_thread; commit:6bf51fdf6afe6f389415792f335dd08adf9fa85a; resolved:0-->
nit: fix indentation of comment